Description: Program vesafb on BIOS systems rather than efifb
 efifb is not appropriate on BIOS systems.  Program vesafb if GRUB already
 set up a VESA mode; otherwise fall back to plain text.
 .
 In future, we may be able to do better once Linux supports a
 hardware-agnostic simple linear framebuffer device.
Author: Colin Watson <cjwatson@ubuntu.com>
Forwarded: http://lists.gnu.org/archive/html/grub-devel/2010-07/msg00036.html
Last-Update: 2010-07-12

Index: b/loader/i386/linux.c
===================================================================
--- a/loader/i386/linux.c
+++ b/loader/i386/linux.c
@@ -463,6 +463,7 @@
   int e820_num;
   grub_err_t err = 0;
   char *modevar, *tmp;
+  grub_video_driver_id_t driver_id;
 
   params = real_mode_mem;
 
@@ -538,13 +539,17 @@
       grub_errno = GRUB_ERR_NONE;
     }
 
+  driver_id = grub_video_get_driver_id ();
+  if (driver_id != GRUB_VIDEO_DRIVER_NONE &&
+      driver_id != GRUB_VIDEO_DRIVER_VBE)
+    /* We currently have no way to pass information about other modes to
+       Linux.  */
+    grub_video_set_mode ("text", 0, 0);
+
   if (! grub_linux_setup_video (params))
     {
-      /* Use generic framebuffer unless VESA is known to be supported.  */
-      if (params->have_vga != GRUB_VIDEO_LINUX_TYPE_VESA)
-	params->have_vga = GRUB_VIDEO_LINUX_TYPE_SIMPLE;
-      else
-	params->lfb_size >>= 16;
+      params->have_vga = GRUB_VIDEO_LINUX_TYPE_VESA;
+      params->lfb_size >>= 16;
     }
   else
     {
